ete4/parser/newick.pyx

```diff
@@ -0,0 +1,172 @@
+
+    if not tree_text.endswith(';'):
+        raise NewickError('text ends with no ";"')
+
+    if tree_text[0] == '(':
+        nodes, pos = read_nodes(tree_text, 0)
+    else:
+        nodes, pos = [], 0
+
+    content, pos = read_content(tree_text, pos)
+    if pos != len(tree_text) - 1:
+        raise NewickError(f'root node ends at position {pos}, before tree ends')
+
+    t = Tree(content)
+    t.children = nodes
+    return  t
+
+
+def read_nodes(nodes_text, int pos=0):
+    "Return a list of nodes and the position in the text where they end"
+    # nodes_text looks like '(a,b,c)', where any element can be a list of nodes
+    if nodes_text[pos] != '(':
+        raise NewickError('nodes text starts with no "("')
+
+    nodes = []
+    while nodes_text[pos] != ')':
+        pos += 1
+        if pos >= len(nodes_text):
+            raise NewickError('nodes text ends missing a matching ")"')
+
+        while nodes_text[pos] in ' \t\r\n':
+            pos += 1  # skip whitespace
+
+        if nodes_text[pos] == '(':
+            children, pos = read_nodes(nodes_text, pos)
+        else:
+            children = []
+
+        content, pos = read_content(nodes_text, pos)
+
+        t = Tree(content)
+        t.children = children
+        nodes.append(t)
+
+    return nodes, pos+1
+
+
+def read_content(str text, int pos, endings=',);'):
+    "Return content starting at position pos in the text, and where it ends"
+    start = pos
+    if pos < len(text) and text[pos] == "'":
+        _, pos = read_quoted_name(text, pos)
+    while pos < len(text) and text[pos] not in endings:
+        pos += 1
+    return text[start:pos], pos
+
+
+def read_quoted_name(str text, int pos):
+    "Return quoted name and the position where it ends"
+    if pos >= len(text) or text[pos] != "'":
+        raise NewickError(f'text at position {pos} does not start with "\'"')
+
+    pos += 1
+    start = pos
+    while pos < len(text):
+        if text[pos] == "'":
+            # Newick format escapes ' as ''
+            if pos+1 >= len(text) or text[pos+1] != "'":
+                return text[start:pos].replace("''", "'"), pos+1
+            pos += 2
+        else:
+            pos += 1
+
+    raise NewickError(f'unfinished quoted name: {text[start:]}')
+
+
+def get_content_fields(content):
+    """Return name, length, properties from the content (as a newick) of a node
+
+    Example:
+      'abc:123[&&NHX:x=foo:y=bar]' -> ('abc', 123, {'x': 'foo', 'y': 'bar'})
+    """
+    cdef double length
+    if content.startswith("'"):
+        name, pos = read_quoted_name(content, 0)
+    else:
+        name, pos = read_content(content, 0, endings=':[')
+
+    if pos < len(content) and content[pos] == ':':
+        length_txt, pos = read_content(content, pos+1, endings='[')
+        try:
+            length = float(length_txt)
+        except ValueError:
+            raise NewickError('invalid number %r in %r' % (length_txt, content))
+    else:
+        length = -1
```

Comment:
Yes, any node if the length the node is not specified. Let me explain.

I would use `length = None` if it was not given, and that's what I used to do, but with cython it is not possible. Since we declare `length` to be a double, it can only be assigned that type. So -1 was playing the role of None for me now, meaning it was not specified (as in many trees, like the [aves](http://treeview.etetoolkit.org/static/gui.html?tree=aves) example, or the [ncbi](http://treeview.etetoolkit.org/static/gui.html?tree=ncbi)), which is not the same as having a 0 size (which happens in many leaves for the [GISAID covid tree](http://treeview.etetoolkit.org/static/gui.html?tree=GISAID) for example).

I'm not sure how ETE differentiates between a node with a length not specified and a node with a length which is truly 0, or if it cares for that difference (and consider trees like the GISAID to be malformed or something)...